src/frontend/src/scheduler/snapshot.rs

```diff
@@ -12,68 +12,84 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use std::collections::HashMap;
+use std::collections::{HashMap, HashSet};
 use std::sync::Arc;
 
-use risingwave_common::util::epoch::Epoch;
+use anyhow::anyhow;
+use risingwave_common::catalog::TableId;
+use risingwave_common::util::epoch::{Epoch, INVALID_EPOCH};
 use risingwave_hummock_sdk::version::HummockVersionStateTableInfo;
 use risingwave_hummock_sdk::{
     FrontendHummockVersion, FrontendHummockVersionDelta, INVALID_VERSION_ID,
 };
 use risingwave_pb::common::{batch_query_epoch, BatchQueryEpoch};
-use risingwave_pb::hummock::{HummockVersionDeltas, PbHummockSnapshot};
+use risingwave_pb::hummock::{HummockVersionDeltas, PbHummockSnapshot, StateTableInfoDelta};
 use tokio::sync::watch;
 
 use crate::expr::InlineNowProcTime;
 use crate::meta_client::FrontendMetaClient;
+use crate::scheduler::SchedulerError;
 
 /// The storage snapshot to read from in a query, which can be freely cloned.
 #[derive(Clone)]
 pub enum ReadSnapshot {
     /// A frontend-pinned snapshot.
     FrontendPinned {
         snapshot: PinnedSnapshotRef,
-        // It's embedded here because we always use it together with snapshot.
-        is_barrier_read: bool,
     },
 
+    ReadUncommitted,
+
     /// Other arbitrary epoch, e.g. user specified.
     /// Availability and consistency of underlying data should be guaranteed accordingly.
     /// Currently it's only used for querying meta snapshot backup.
     Other(Epoch),
 }
 
-impl ReadSnapshot {
+pub struct QuerySnapshot {
```

Comment:
Can we have more documentation on this struct?

Reply:
Just refactor the code, and the new `QuerySnapshot` is removed.